// @coyoapp/plugin-adapter/build/main/lib/plugin-adapter.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const jest_createspyobj_1 = require("jest-createspyobj");
const utils_1 = require("ts-jest/utils");
const config_adapter_1 = require("./adapter/config-adapter");
const height_adapter_1 = require("./adapter/height-adapter");
const init_adapter_1 = require("./adapter/init-adapter");
const jwk_defaults_1 = require("./jwt/jwk-defaults");
const jwk_store_1 = require("./jwt/jwk-store");
const plugin_adapter_1 = require("./plugin-adapter");
jest.mock('./jwt/jwk-store');
jest.mock('./adapter/init-adapter');
jest.mock('./adapter/config-adapter');
jest.mock('./adapter/height-adapter');
describe('PluginAdapter', () => {
    let windowSpy;
    let jwkStoreMock;
    let initAdapterMock;
    let configAdapterMock;
    let heightAdapterMock;
    let pluginAdapter;
    beforeEach(() => {
        windowSpy = jest.spyOn(window, 'window', 'get');
        utils_1.mocked(jwk_store_1.JwkStore).getInstance.mockImplementation(() => jwkStoreMock);
        jwkStoreMock = jest_createspyobj_1.createSpyObj(jwk_store_1.JwkStore);
        utils_1.mocked(init_adapter_1.InitAdapter).mockImplementation(() => initAdapterMock);
        initAdapterMock = jest_createspyobj_1.createSpyObj(init_adapter_1.InitAdapter);
        utils_1.mocked(config_adapter_1.ConfigAdapter).mockImplementation(() => configAdapterMock);
        configAdapterMock = jest_createspyobj_1.createSpyObj(config_adapter_1.ConfigAdapter);
        utils_1.mocked(height_adapter_1.HeightAdapter).mockImplementation(() => heightAdapterMock);
        heightAdapterMock = jest_createspyobj_1.createSpyObj(height_adapter_1.HeightAdapter);
        pluginAdapter = new plugin_adapter_1.PluginAdapter('srcId');
    });
    afterEach(() => {
        windowSpy.mockRestore();
    });
    test('should use the given src ID', () => {
        expect(pluginAdapter.srcId).toBe('srcId');
    });
    test('should use the query src ID', () => {
        windowSpy.mockImplementation(() => ({
            location: {
                href: 'https://example.com?src=querySrcId'
            }
        }));
        pluginAdapter = new plugin_adapter_1.PluginAdapter();
        expect(pluginAdapter.srcId).toBe('querySrcId');
    });
    test('should fallback without any src ID', () => {
        pluginAdapter = new plugin_adapter_1.PluginAdapter();
        expect(pluginAdapter.srcId).toBe('');
    });
    test('should register hosts', () => {
        expect(jwkStoreMock.addHost.mock.calls).toContainEqual(['https://certificates.plugins.coyoapp.com']);
    });
    test('should register keys', () => {
        expect(jwkStoreMock.addJwk.mock.calls).toContainEqual([jwk_defaults_1.JWK_STAGE_RGX, jwk_defaults_1.JWK_STAGE, jwk_defaults_1.JWK_STAGE_NEW]);
        expect(jwkStoreMock.addJwk.mock.calls).toContainEqual([jwk_defaults_1.JWK_PROD_RGX, jwk_defaults_1.JWK_PROD, jwk_defaults_1.JWK_PROD_NEW]);
    });
    test('should init', () => {
        const claims = {
            sub: 'subject',
            'ctx.userId': 'userID',
            'cfg.text': 'Hello World',
            'cfg.bool': true,
            'stt.edit': false
        };
        const token = 'token';
        initAdapterMock.send.mockReturnValueOnce(Promise.resolve([claims, token]));
        const result = pluginAdapter.init();
        return result.then(v => {
            expect(initAdapterMock.send).toBeCalledWith(60000, true);
            expect(v).toStrictEqual({
                claims: {
                    cfg: {
                        bool: true,
                        text: 'Hello World'
                    },
                    ctx: {
                        userId: 'userID'
                    },
                    stt: {
                        edit: false
                    },
                    sub: 'subject'
                },
                token: 'token',
                validated: true
            });
        });
    });
    test('should connect to config adapter', () => {
        const cb = () => {
            /* empty */
        };
        const result = pluginAdapter.onSave(cb);
        expect(configAdapterMock.connect).toHaveBeenCalledWith(cb, true);
        result();
        expect(configAdapterMock.disconnect).toHaveBeenCalled();
    });
    test('should connect to height adapter', () => {
        const result = pluginAdapter.observeHeight();
        expect(heightAdapterMock.connect).toHaveBeenCalledWith('html', 50);
        result();
        expect(heightAdapterMock.disconnect).toHaveBeenCalled();
    });
});
