// @coyoapp/plugin-adapter/build/main/lib/plugin-adapter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PluginAdapter = void 0;
const config_adapter_1 = require("./adapter/config-adapter");
const height_adapter_1 = require("./adapter/height-adapter");
const init_adapter_1 = require("./adapter/init-adapter");
const jwk_defaults_1 = require("./jwt/jwk-defaults");
const jwk_store_1 = require("./jwt/jwk-store");
const nest_1 = require("./utils/nest");
/**
 * The main service class to be used by a plug-in to establish a safe and secure
 * communication to the COYO front end bridge. It acts as a unified interaction
 * interface to all other plug-in handlers.
 */
class PluginAdapter {
    /**
     * Creates a new instance of this class.
     *
     * @param srcId the plug-in source ID to use for requests
     * @returns the new instance
     */
    constructor(srcId) {
        this.srcId = srcId || new URL(window.location.href).searchParams.get('src') || '';
        this.jwkStore = jwk_store_1.JwkStore.getInstance();
        this.initAdapter = new init_adapter_1.InitAdapter(this.srcId);
        this.initConfigAdapter = new init_adapter_1.InitAdapter(this.srcId + PluginAdapter.CONFIG_SRC_SUFFIX);
        this.heightAdapter = new height_adapter_1.HeightAdapter(this.srcId);
        this.heightConfigAdapter = new height_adapter_1.HeightAdapter(this.srcId + PluginAdapter.CONFIG_SRC_SUFFIX);
        this.configAdapter = new config_adapter_1.ConfigAdapter(this.srcId + PluginAdapter.CONFIG_SRC_SUFFIX);
        this.jwkStore.addHost('https://certificates.plugins.coyoapp.com');
        this.jwkStore.addJwk(jwk_defaults_1.JWK_STAGE_RGX, jwk_defaults_1.JWK_STAGE, jwk_defaults_1.JWK_STAGE_NEW);
        this.jwkStore.addJwk(jwk_defaults_1.JWK_PROD_RGX, jwk_defaults_1.JWK_PROD, jwk_defaults_1.JWK_PROD_NEW);
    }
    /**
     * Sends an initialization message to the parent window (i.e. COYO)
     * and retrieves the decoded (and possibly validated) response as well as the
     * raw token response.
     *
     * @param config a flag indicating if the plugin configuration view should be initialized
     * @param timeout a maximum time to wait for the response
     * @param validate a flag indicating if the JWT response should be validated
     * @returns a promise holding the response data
     */
    init(config = false, timeout = 60000, validate = true) {
        const adapter = config ? this.initConfigAdapter : this.initAdapter;
        return adapter.send(timeout, validate).then(([claims, token]) => {
            if (config) {
                this.configAdapter.bind(claims['ctx.userId']);
            }
            return { claims: nest_1.nest(claims), token, validated: validate };
        });
    }
    /**
     * Starts to listen to COYO requests to save the current configuration.
     * Provides a `respond` callback that is used to respond to the received
     * message. The response can be `true` to indicate success without providing
     * any data, `false` to indicate failure (e.g. an invalid configuration state)
     * or a JSON object holding the configuration values to be stored on COYO
     * side.
     *
     * @param cb a callback that is executed every time a save request is received
     * @param validate a flag indicating if the JWT response should be validated
     * @returns a disconnect handler to stop listening to save events
     */
    onSave(cb, validate = true) {
        this.configAdapter.connect(cb, validate);
        return () => this.configAdapter.disconnect();
    }
    /**
     * Starts to observe the HTML element that matches the given selectors and
     * sends throttled update messages to COYO.
     *
     * @param config a flag indicating if the plugin configuration view should be observed
     * @param selectors the selectors to target the HTML element
     * @returns a disconnect handler to stop observing the HTML element
     */
    observeHeight(config = false, selectors = 'html') {
        const adapter = config ? this.heightConfigAdapter : this.heightAdapter;
        adapter.connect(selectors, 50);
        return () => adapter.disconnect();
    }
}
exports.PluginAdapter = PluginAdapter;
PluginAdapter.CONFIG_SRC_SUFFIX = '-cfg';
