// @coyoapp/plugin-adapter/build/module/lib/adapter/config-adapter.js
import { showError, showErrorAndRecover } from '../error/error-handlers';
import { JwkStore } from '../jwt/jwk-store';
import { JwtHandler } from '../jwt/jwt-handler';
import { randomStr } from '../utils/random';
/**
 * Service class to react to changes in the config form on COYO side.
 */
export class ConfigAdapter {
    /**
     * Creates a new instance of this class.
     *
     * @param srcId the plug-in source ID to use for requests
     * @returns the new instance
     */
    constructor(srcId) {
        this.srcId = srcId;
        this.listener = null;
        this.userId = null;
        this.jwkStore = JwkStore.getInstance();
        this.jwtHandler = new JwtHandler();
    }
    /**
     * Bind this instance to a specific user ID to only accept tokens that define
     * the same user ID via "ctx.userId".
     *
     * @param userId the user ID
     */
    bind(userId) {
        this.userId = userId;
    }
    /**
     * Connects the given callback listeners to config form events coming from COYO.
     *
     * @param cb a callback to be executed when the form should be saved
     * @param validate a flag indicating if the JWT response should be validated
     */
    connect(cb, validate) {
        this.disconnect();
        this.listener = (event) => {
            const claims = this.jwtHandler.decodeJwtClaims(event.data);
            if (claims && claims['sub'] === ConfigAdapter.REQ_SUBJECT && claims['ctx.userId'] === this.userId) {
                const jti = claims.jti || randomStr();
                const respond = (data) => parent.postMessage({ iss: this.srcId, sub: ConfigAdapter.RES_SUBJECT, jti, data }, '*');
                if (validate) {
                    return this.jwtHandler
                        .validateJwt(claims, {})
                        .then(() => {
                        const header = this.jwtHandler.decodeJwtHeader(event.data);
                        return this.jwkStore
                            .getJwk(claims.iss || '', header ? header.jku : undefined)
                            .then(keys => this.jwtHandler.verifyJwt(event.data, ...keys))
                            .catch(err => (err.recover ? showErrorAndRecover(err, void 0) : Promise.reject(err)));
                    })
                        .then(() => cb(respond, event.data))
                        .catch(err => showError(err));
                }
                else {
                    cb(respond, event.data);
                }
            }
            return Promise.resolve();
        };
        window.addEventListener('message', this.listener);
    }
    /**
     * Disconnects all registered listeners.
     */
    disconnect() {
        if (this.listener) {
            window.removeEventListener('message', this.listener);
            this.listener = null;
        }
    }
}
ConfigAdapter.REQ_SUBJECT = 'save';
ConfigAdapter.RES_SUBJECT = 'setConfig';
