// @coyoapp/plugin-adapter/build/main/lib/jwt/jwk-defaults.js
"use strict";
// -----
// - JSON Web Keys (JWKs) generated via https://russelldavies.github.io/jwk-creator/
// -----
Object.defineProperty(exports, "__esModule", { value: true });
exports.JWK_PROD_NEW = exports.JWK_PROD = exports.JWK_PROD_RGX = exports.JWK_STAGE_NEW = exports.JWK_STAGE = exports.JWK_STAGE_RGX = void 0;
/** @ignore */
exports.JWK_STAGE_RGX = /^https?:\/\/(?!next(?:-pmm)?\.coyostaging\.com)[a-z0-9_-]+\.coyostaging\.com/i;
/** @ignore */
exports.JWK_STAGE = {
    alg: {
        name: 'RSASSA-PKCS1-v1_5',
        hash: { name: 'SHA-512' }
    },
    key: {
        kty: 'RSA',
        n: '65n5VlWdmtiJFJHD-W9swkx4GQvZHI9RN1L2Ku0qKK3oOOeedETA8PAb3xOGN6g9frxZ29UfgVJ4DqPN' +
            'xkJHMLqdoqrCmIArTBYp7jSmEURxXnYdNQq4NKVrUgT7cvuiRU1QHUxKJkKGZuywzHDCWHCzPWaq__ZZ' +
            'D-PV7CG-LNwVqDiIt1GyDBTnBSuhJT3Wh7xD8qlMxflYjqYuVAmXRtT4ZxoZz6Swvoun1r5_FDKPoite' +
            'dRez3xQeFvBvPVE23d5jzesd335Jogtxn1JXEJErrg3NvdYudcH7EUJzrIiBwXhQlXfsSE0oit0ukJmN' +
            'DZaYyFkIrWE23dlvGVGkBTl5kfGLTNB9VVPdfPl2BYxbk_sVitWTqNQgyfOJBHC0_Szf0HSMmOeNY9z3' +
            'AxGQhhriNRONwoQLZiw48YTP9EdbktnwrpsQ9BnlDKZHvK_tCAX02pm7sIAJvxdGOub1l409x2dZnNmb' +
            'xXQzDJV8mvQKprfNiKt_ewL6NoxHSjLwoVwmh6kCdXvp6yv4F9nc0A6nuZoyeV7N9S1ViF7kPeZ5HbGu' +
            'EjuKB90LEqHz-QqHo2D6qjUumbQxqsiboizw1V-dUrJrJzNxviqoYQywKw7gcluyJQ274rtMD4C1QhXZ' +
            'nf_d9NLD69pi1nD2HWluQ_DAMBzRJtE04WkzQXASGTc',
        e: 'AQAB',
        alg: 'RS512',
        use: 'sig'
    }
};
/** @ignore */
exports.JWK_STAGE_NEW = {
    alg: {
        name: 'RSASSA-PKCS1-v1_5',
        hash: { name: 'SHA-512' }
    },
    key: {
        kty: 'RSA',
        n: 'wk_n-oxXDFUrBKpZsicqAF7ECC_FBjXjY9btUmewaKN09JMZ7kpXoh1SMPmtphlMhwsaqWmr9bRXu2d6' +
            '5prGdA8gkoCCjzjR7YpDGdr4OTWBfxzvgJG8-5gMa-DRfyZjC__64oz1jAKXcetT10I38-zmZeFNFp82' +
            'RFaHNmgcnY7qt_FdWY-4JUFKvQLwXn8pYBougEEHVQW1BPmJbxiHBbe0L0T9-qVHZEWzm39_EyhK2wPD' +
            'rC4Y6j4Deb70K5zHnRFRgcLTw9TUGvf2l4QjIL-U_pF8CneIKi-vUHgOr1Kv0a0kik59omV_7QBQDa9Z' +
            'utRAGgWcb-sOr6b8Qeay_lpZRHTmiHORLQPJrF-j8jeaDEK9wm5SmtsjD-UGHEGlEtnHV9XPI6dYgta7' +
            'kR33JDCyJcGA2ba6KrETVnToMw5qJoT2WxS4xBfMkVK-k-NdQ4pmgrwEg0lphdylrKUct_1HcmtBNKVT' +
            'lzbwfpj0vvFqWFRf4lD9i5aZabkHQY82kp21k-hl5zk7htTHq63Khf3NVzU9EzEW56njRkxN6XsapN-a' +
            'fCLlNRK3StbCYod5Y3vhqfyuxZSinb-HKcL5AsoxqmsHEsWZa5N724KWyS7J5vqDymKsQGxFFXDbiAVU' +
            'gu5_6BjccMevj7RLkGhXqDHOp8jZ4mUjZgxCwJ1LziM',
        e: 'AQAB',
        alg: 'RS512',
        use: 'sig'
    }
};
/** @ignore */
exports.JWK_PROD_RGX = /^https?:\/\/[a-z0-9_-]+\.coyocloud.com/i;
/** @ignore */
exports.JWK_PROD = {
    alg: {
        name: 'RSASSA-PKCS1-v1_5',
        hash: { name: 'SHA-512' }
    },
    key: {
        kty: 'RSA',
        n: '3JJaLM6VoIzAZJXxCGOtJheDnDAqD9Bz3UTdvTqh84HJVm8eYqxJpCm6x9eFuj5FfZ486Jwtx145A47f' +
            'ds9wbi646BJ2PfBxWv1jB9ST70hdatq9zEglUN3L7Dm1yXnJlwqgTtGCuJvNCxoTjNa1RvSSWr9xQ7JS' +
            'Agt1xmtwaTgJDdASdxsFTo0A-51l5vAU6pF6lheQ3WJYNPOY0k4S07jTnuVIeN4uUC6dt0xQpxynmumr' +
            'Cez-o4EOL-TFunUGsJHpi9jB31W7QFMLUBzsDNJDcEzH5VfilZsDy81ZC068SiPqEiUIMNT06XjZrwNe' +
            'N5HFi_5qCUogAVUPma31OEOxjNotWnuJZkK93j1aJ3HdHTCLK_pF_odCrKCKrN14He-EZY1h04pat7PV' +
            'L_w53eMquHwb34J5tgDlsr3roQB61TRDaswx9HgexEjLWvcBk-nPAtpyRfiT6Ytjuy8Nds76qFmJieP3' +
            '_QDIRjJQdZSExTL_jMfax0mmzr9lEsKD5Iq237Z2vVu6wiJyCGwJYc00KKKo8Ux_BzuBtwoI59F7txUX' +
            'FIBOK09l9yoq8yCx0qMArnd7bFrov3kSVTfo-QHxBZ2tKy1My4qOmByK5M1G1TKjIQT-sEWKbmI8d0sM' +
            '2Nk3yRH_6b_GB7mWTQsw7lXpeMGfEC8p7ws2KBU6810',
        e: 'AQAB',
        alg: 'RS512',
        use: 'sig'
    }
};
/** @ignore */
exports.JWK_PROD_NEW = {
    alg: {
        name: 'RSASSA-PKCS1-v1_5',
        hash: { name: 'SHA-512' }
    },
    key: {
        kty: 'RSA',
        n: 'wLXTeg79fOGfkpFhDVhClPtSIP24gO-s6cc4R_NtKP4iwYuLamcwlvlmeoUU5XbHvVyazZBRGnSedXwe' +
            '2gSClDm-Vb9Ar6DyoTBYbtaHj5JJI_pDNJnct_BtdqzdQGxVNLAMn6hUVaqHPnqKVjNHLotf-IadJ1Ms' +
            'ewKwTTb7dxEIQnk1JoSS7PQtrsoXXfcBsUgX5zwS3EH6uc2zeW4e0BRcOnyTLuYYdczUL8qtHNI7zFKQ' +
            'KAIwqNaQUQ9ekbBkbtcNejidysJ0bOqjGxtCZTt7Ib3xyM2Oe2uAV2VE8zUdCfRHbxjDJc0fcCkUalz1' +
            'AKgm8ZBDEo4izIlM71O_gqoXBq7nUxe9BDGVccxr16uuneGfq8aTxGE2zBSnevJTxktFq0gCnRFgVm_q' +
            'Kb4_3Mr4AKrbGShXlbzrGs8bm1_EfGWlLIxbSF2gDaUSJfLVQSPmQmPHQnjHFM32MPr09w1LhCuUSZBi' +
            'sG9t86zPjiEvZ0fjpOudM6QvmJFETqnpeSCi3fjQA2F0e-mSJ0N8HMlIW8Wj-Nmit1f1MJ2o5uEdj_b1' +
            'PNcWu2AMdFzt2_ak1DcXyi8u26kyDlu8ZM4zIajhXNDwx2m7CvsYd-1mZCjgxWxfYKqDLVXEzOp6dcx4' +
            'g8ShXJYydSd5liNyH7d_LS_WK5qHoQAIvneMOsKIWus',
        e: 'AQAB',
        alg: 'RS512',
        use: 'sig'
    }
};
