// @coyoapp/plugin-adapter/build/main/lib/adapter/config-adapter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConfigAdapter = void 0;
const error_handlers_1 = require("../error/error-handlers");
const jwk_store_1 = require("../jwt/jwk-store");
const jwt_handler_1 = require("../jwt/jwt-handler");
const random_1 = require("../utils/random");
/**
 * Service class to react to changes in the config form on COYO side.
 */
class ConfigAdapter {
    /**
     * Creates a new instance of this class.
     *
     * @param srcId the plug-in source ID to use for requests
     * @returns the new instance
     */
    constructor(srcId) {
        this.srcId = srcId;
        this.listener = null;
        this.userId = null;
        this.jwkStore = jwk_store_1.JwkStore.getInstance();
        this.jwtHandler = new jwt_handler_1.JwtHandler();
    }
    /**
     * Bind this instance to a specific user ID to only accept tokens that define
     * the same user ID via "ctx.userId".
     *
     * @param userId the user ID
     */
    bind(userId) {
        this.userId = userId;
    }
    /**
     * Connects the given callback listeners to config form events coming from COYO.
     *
     * @param cb a callback to be executed when the form should be saved
     * @param validate a flag indicating if the JWT response should be validated
     */
    connect(cb, validate) {
        this.disconnect();
        this.listener = (event) => {
            const claims = this.jwtHandler.decodeJwtClaims(event.data);
            if (claims && claims['sub'] === ConfigAdapter.REQ_SUBJECT && claims['ctx.userId'] === this.userId) {
                const jti = claims.jti || random_1.randomStr();
                const respond = (data) => parent.postMessage({ iss: this.srcId, sub: ConfigAdapter.RES_SUBJECT, jti, data }, '*');
                if (validate) {
                    return this.jwtHandler
                        .validateJwt(claims, {})
                        .then(() => {
                        const header = this.jwtHandler.decodeJwtHeader(event.data);
                        return this.jwkStore
                            .getJwk(claims.iss || '', header ? header.jku : undefined)
                            .then(keys => this.jwtHandler.verifyJwt(event.data, ...keys))
                            .catch(err => (err.recover ? error_handlers_1.showErrorAndRecover(err, void 0) : Promise.reject(err)));
                    })
                        .then(() => cb(respond, event.data))
                        .catch(err => error_handlers_1.showError(err));
                }
                else {
                    cb(respond, event.data);
                }
            }
            return Promise.resolve();
        };
        window.addEventListener('message', this.listener);
    }
    /**
     * Disconnects all registered listeners.
     */
    disconnect() {
        if (this.listener) {
            window.removeEventListener('message', this.listener);
            this.listener = null;
        }
    }
}
exports.ConfigAdapter = ConfigAdapter;
ConfigAdapter.REQ_SUBJECT = 'save';
ConfigAdapter.RES_SUBJECT = 'setConfig';
