// @coyoapp/plugin-adapter/build/main/lib/adapter/init-adapter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InitAdapter = void 0;
const error_handlers_1 = require("../error/error-handlers");
const plugin_error_1 = require("../error/plugin-error");
const jwk_store_1 = require("../jwt/jwk-store");
const jwt_handler_1 = require("../jwt/jwt-handler");
const random_1 = require("../utils/random");
/**
 * Service class to initialize a plug-in and retrieve initialization response
 * information.
 */
class InitAdapter {
    /**
     * Creates a new instance of this class.
     *
     * @param srcId the plug-in source ID to use for requests
     * @returns the new instance
     */
    constructor(srcId) {
        this.srcId = srcId;
        this.jwkStore = jwk_store_1.JwkStore.getInstance();
        this.jwtHandler = new jwt_handler_1.JwtHandler();
    }
    /**
     * Sends an initialization message to the parent window (i.e. COYO)
     * and retrieves the response.
     *
     * @param timeout a maximum time to wait for the response
     * @param validate a flag indicating if the JWT response should be validated
     * @returns a promise holding the decoded (and possibly validated) JWT claims
     */
    send(timeout, validate) {
        return new Promise((resolve, reject) => {
            let listener = null;
            let timer = null;
            const clear = () => {
                if (listener !== null) {
                    window.removeEventListener('message', listener);
                }
                if (timer !== null) {
                    clearTimeout(timer);
                }
            };
            const jti = random_1.randomStr();
            listener = (event) => {
                const claims = this.jwtHandler.decodeJwtClaims(event.data);
                if (claims && claims.jti === jti) {
                    clear();
                    if (validate) {
                        this.jwtHandler
                            .validateJwt(claims, { sub: InitAdapter.REQ_SUBJECT })
                            .then(() => {
                            const header = this.jwtHandler.decodeJwtHeader(event.data);
                            return this.jwkStore
                                .getJwk(claims.iss || '', header ? header.jku : undefined)
                                .then(keys => this.jwtHandler.verifyJwt(event.data, ...keys))
                                .catch(err => (err.recover ? error_handlers_1.showErrorAndRecover(err, void 0) : Promise.reject(err)));
                        })
                            .then(() => resolve([claims, event.data]))
                            .catch(err => error_handlers_1.showError(err));
                    }
                    else {
                        resolve([claims, event.data]);
                    }
                }
            };
            timer = window.setTimeout(() => {
                clear();
                reject(new plugin_error_1.PluginError(plugin_error_1.PluginErrorCode.MessageTimeout, `Initialization timed out after ${timeout}ms`));
            }, timeout);
            window.addEventListener('message', listener);
            parent.postMessage({ iss: this.srcId, sub: InitAdapter.RES_SUBJECT, jti }, '*');
        });
    }
}
exports.InitAdapter = InitAdapter;
InitAdapter.REQ_SUBJECT = 'init';
InitAdapter.RES_SUBJECT = 'init';
