// @coyoapp/plugin-adapter/build/module/lib/error/plugin-error.js
/**
 * A list of error codes for plug-in related errors.
 */
export var PluginErrorCode;
(function (PluginErrorCode) {
    /**
     * The SubtleCrypto interface of the Web Crypto API is not available.
     */
    PluginErrorCode[PluginErrorCode["CryptoNotFound"] = 0] = "CryptoNotFound";
    /**
     * No CryptoKey could be found for the corresponding environment URL in the JWK store.
     */
    PluginErrorCode[PluginErrorCode["CryptoKeyNotFound"] = 1] = "CryptoKeyNotFound";
    /**
     * The token signature is invalid.
     */
    PluginErrorCode[PluginErrorCode["InvalidTokenSignature"] = 2] = "InvalidTokenSignature";
    /**
     * A token claim is invalid.
     */
    PluginErrorCode[PluginErrorCode["InvalidTokenClaim"] = 3] = "InvalidTokenClaim";
    /**
     * The token is expired.
     */
    PluginErrorCode[PluginErrorCode["InvalidTokenExp"] = 4] = "InvalidTokenExp";
    /**
     * The token is premature.
     */
    PluginErrorCode[PluginErrorCode["InvalidTokenNbf"] = 5] = "InvalidTokenNbf";
    /**
     * The message request timed out.
     */
    PluginErrorCode[PluginErrorCode["MessageTimeout"] = 6] = "MessageTimeout";
    /**
     * No CryptoKey could be found at the given URL.
     */
    PluginErrorCode[PluginErrorCode["CryptoKeyNotFoundRemote"] = 7] = "CryptoKeyNotFoundRemote";
    /**
     * URL for CryptoKey not trusted.
     */
    PluginErrorCode[PluginErrorCode["CryptoKeyNotTrusted"] = 8] = "CryptoKeyNotTrusted";
})(PluginErrorCode || (PluginErrorCode = {}));
/**
 * A plug-in related error thrown in the context of this adapter.
 */
export class PluginError extends Error {
    /**
     * Creates a new instance of this class.
     *
     * @param code the error code
     * @param message the error message
     * @param recover whether the app can recover from the error
     * @returns the new instance
     */
    constructor(code, message, recover = false) {
        super(`${message} (ERR${code})`);
        this.code = code;
        this.recover = recover;
        this.name = 'PluginError';
    }
}
